build-tools/packages/build-infrastructure/src/package.ts

```diff
@@ -0,0 +1,292 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { existsSync } from "node:fs";
+import path from "node:path";
+
+// Imports are written this way for CJS/ESM compat
+import fsePkg from "fs-extra";
+const { readJsonSync } = fsePkg;
+import colors from "picocolors";
+
+import { type WorkspaceDefinition, findReleaseGroupForPackage } from "./config.js";
+import { readPackageJsonAndIndent, writePackageJson } from "./packageJsonUtils.js";
+import type {
+	AdditionalPackageProps,
+	IPackage,
+	IPackageManager,
+	PackageDependency,
+	PackageJson,
+	PackageName,
+	ReleaseGroupName,
+} from "./types.js";
+import { lookUpDirSync } from "./utils.js";
+
+export abstract class PackageBase<
+	TAddProps extends AdditionalPackageProps = undefined,
+	J extends PackageJson = PackageJson,
+> implements IPackage
+{
+	// eslint-disable-next-line @typescript-eslint/prefer-readonly -- false positive; this value is changed
+	private static packageCount: number = 0;
+	private static readonly colorFunction = [
+		colors.red,
+		colors.green,
+		colors.yellow,
+		colors.blue,
+		colors.magenta,
+		colors.cyan,
+		colors.white,
+		colors.gray,
+		colors.redBright,
+		colors.greenBright,
+		colors.yellowBright,
+		colors.blueBright,
+		colors.magentaBright,
+		colors.cyanBright,
+		colors.whiteBright,
+	];
+
+	private readonly _indent: string;
+	private _packageJson: J;
+	private readonly packageId = Package.packageCount++;
+
+	// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+	private get color() {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		return Package.colorFunction[this.packageId % Package.colorFunction.length]!;
+	}
+
+	/**
+	 * Create a new package from a package.json file. **Prefer the .load method to calling the contructor directly.**
+	 *
+	 * @param packageJsonFilePath - The path to a package.json file.
+	 * @param packageManager - The package manager used by the workspace.
+	 * @param isWorkspaceRoot - Set to true if this package is the root of a workspace.
+	 * @param additionalProperties - An object with additional properties that should be added to the class. This is
+	 * useful to augment the package class with additional properties.
```

Comment:
These were used in the previous version of build-tools, but I don't think they're needed now.